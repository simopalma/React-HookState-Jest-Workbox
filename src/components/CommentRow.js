import React from 'react';


function CommentRow(props) {
    const renderAlternateComment = () => {
        let html;

        if (props.direction === 'left') {
            html = (
                <section className="message -left">
                    <i className="nes-bcrikko icon" />
                    <div className="nes-balloon from-left">
                        <p>{props.comment.body}</p>
                    </div>
                </section>  
            );
        } else {
            html = (
                <section className="message -right">
                    <div className="nes-balloon from-right">
                        <p>{props.comment.body}</p>
                    </div>
                    <i className="nes-bcrikko icon" />
                </section>
            );
        }

        return html;
    }
    return renderAlternateComment();
}

export default CommentRow;
