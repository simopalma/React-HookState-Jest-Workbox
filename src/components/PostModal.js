import React from 'react';

function PostModal(props) {
    const renderUserCard = () => {
        let html = <span />;

        if (props.detailedUser.name) {
            html = (
                <div className="modal-container">
                    <i className="nes-ash" />
                    <p>{props.detailedUser.name}</p>
                    <p>{props.detailedUser.email}</p>
                    <p>{props.detailedUser.phone}</p>
                    <p>{props.detailedUser.website}</p>
                    <div className="details-container">
                        <span className="details-title"> Company </span>
                        <span className="container-info">{props.detailedUser.company.name}</span>
                        <span className="container-info">{props.detailedUser.company.catchPhrase}</span>
                        <span className="container-info">{props.detailedUser.company.bs}</span>
                    </div>
                    <div className="details-container">
                        <span className="details-title"> Address </span>
                        <span className="container-info">{props.detailedUser.address.street}</span>
                        <span className="container-info">{props.detailedUser.address.suite}</span>
                        <span className="container-info">{props.detailedUser.address.city}</span>
                        <span className="container-info">{props.detailedUser.address.zipcode}</span>
                    </div>
                </div>
            );
        }

        return html;
    }

    return (
        <dialog className="nes-dialog is-rounded" id="modal-post" onClick={props.closeModal}>
            {renderUserCard()}
        </dialog>
    );
}

export default PostModal;
