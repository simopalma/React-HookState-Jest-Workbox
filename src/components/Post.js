import React, { useState } from 'react';
import axios from 'axios';
import CommentRow from './CommentRow';
import constants from '../constants';

function Post(props) {
    const [comments, setComments] = useState([]);

    const loadComments = () => {
        axios.get(`${constants.jsonCommentsUrl}?postId=${props.post.id}`)
            .then(response => {
                setComments(response.data);
            }
        );
    }

    const renderComments = () => {
        let html;

        if (comments.length > 0) {
            html = (
                <section className="message-list">
                    {
                        comments.map((comment, index) => {
                            const direction = index % 2 ? 'right' : 'left';
                            return <CommentRow key={comment.id} comment={comment} direction={direction} />;
                        })
                    }
                </section>
            );
        } else {
            html = (
                <span className="comments-link nes-pointer" onClick={loadComments}> Load Comments </span>
            );
        }

        return html;
    }

    return (
        <div className="post-container">
            <p className="user nes-pointer" onClick={() => props.openModal(props.findUser(props.post.userId))}>
                <i className="nes-ash" /> <span> {props.findUser(props.post.userId).name} </span>
            </p>
            <p className="title"> {props.post.title} </p>
            <p className="body"> {props.post.body} </p>
            {renderComments()}
        </div>
    );
}

export default Post;
