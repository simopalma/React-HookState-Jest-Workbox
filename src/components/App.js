import React, { useState } from 'react';
import Post from './Post';
import PostModal from './PostModal';
import constants from '../constants';
import axios from 'axios';

function App() {
    const modalElement = document.querySelector('#modal-post');
    const [posts, setPosts] = useState([]);
    const [detailedUser, setdetailedUser] = useState({});
    const [users, setUsers] = useState([]);

    if (users.length <= 0) {
        axios.all([
            axios.get(constants.jsonUsersUrl),
            axios.get(constants.jsonPostUrl)
        ]).then(axios.spread((users, posts) => {
            setUsers(users.data);
            setPosts(posts.data);
        }));
    }

    const findUser = id => {
        return users.find(user => user.id === id);
    };

    const openModal = user => {
        setdetailedUser(user);
        modalElement.showModal();
    }

    const closeModal = (e) => {
        if (modalElement.attributes.open !== undefined && !e.target.closest('.modal-container')) {
            modalElement.close();
        }
    }

    return (
        <div>
            <div className="app-container">
                {posts.map(post => (
                    <Post key={post.id} post={post} openModal={openModal} findUser={findUser} />
                ))}
            </div>
            <PostModal closeModal={closeModal} detailedUser={detailedUser} />
        </div>
    );
}

export default App;
