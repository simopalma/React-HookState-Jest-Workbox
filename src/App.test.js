import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import App from './components/App';
import Post from './components/Post';
import CommentRow from './components/CommentRow';
import PostModal from './components/PostModal';

describe('React Test', () => {
    let wrapper;
    describe('App', () => {
        beforeEach(() => {
            wrapper = Enzyme.shallow(<App />);
        });

        it('Should render correctly', () => {
            const tree = renderer.create(<App />);
            expect(tree.toJSON()).toMatchSnapshot();
        });
    });
    describe('PostModal', () => {
        const props = {
            detailedUser: {},
            closeModal: jest.fn()
        };

        beforeEach(() => {
            wrapper = Enzyme.shallow(<PostModal {...props} />);
        });

        it('Should render correctly', () => {
            const tree = renderer.create(<PostModal {...props} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });
        it('Should call close modal function on click outside', () => {
            const modalTrigger = wrapper.find('#modal-post');
            modalTrigger.simulate('click');
            expect(props.closeModal).toHaveBeenCalled();
        });
        it('Should not render user detail if not passed', () => {
            const modalContainer = wrapper.find('.modal-container');
            expect(modalContainer.length).not.toBeGreaterThan(0);
        });
        it('Should render user detail when passed', () => {
            props.detailedUser = {
                name: 'test name',
                company: {},
                address: {},
            }
            wrapper = Enzyme.shallow(<PostModal {...props} />);
            const modalContainer = wrapper.find('.modal-container');
            expect(modalContainer.length).toBeGreaterThan(0);
        });
    });
    describe('Post', () => {
        const props = {
            post: {
                userId: 1,
                id: 1,
                title: 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
                body: 'lorem',
                comments: [{id: 1, body: 'test'}],
            },
            findUser: id => { return {name: 'test name'} },
            openModal: jest.fn(),
        };

        beforeEach(() => {
            wrapper = Enzyme.shallow(<Post {...props} />);
        });

        it('Should render correctly', () => {
            const tree = renderer.create(<Post {...props} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });
        it('Call open modal on click of the user', () => {
            const userPointer = wrapper.find('.user');
            userPointer.simulate('click');
            expect(props.openModal).toHaveBeenCalled();
        });
        it('Comments load function called on click', () => {
            const commentsPointer = wrapper.find('.comments-link');
            commentsPointer.simulate('click');

            // Simulate Ajax call of the comments
            setTimeout(() => {
                expect(wrapper.find('.message-list').length).toBeGreaterThan(0);
            });
        });
    });
    describe('CommentRow', () => {
        const props = {
            comment: [{ id: 1, body: '' }, { id: 2, body: '' }],
            direction: 'left',
        };

        beforeEach(() => {
            wrapper = Enzyme.shallow(<CommentRow {...props} />);
        });

        it('Should render correctly', () => {
            const tree = renderer.create(<CommentRow {...props} />);
            expect(tree.toJSON()).toMatchSnapshot();
        });
        it('Should render left comment', () => {
            const leftComment = wrapper.find('.-left');
            // const rightComment = wrapper.find('.-right');
            expect(leftComment.length).toBeGreaterThan(0);
        });
        it('Should render right comment', () => {
            props.direction = 'right';
            wrapper = Enzyme.shallow(<CommentRow {...props} />);
            const rightComment = wrapper.find('.-right');
            expect(rightComment.length).toBeGreaterThan(0);
        });
    });
});
