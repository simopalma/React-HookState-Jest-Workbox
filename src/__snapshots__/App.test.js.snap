// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Test App Should render correctly 1`] = `
<div>
  <div
    className="app-container"
  />
  <dialog
    className="nes-dialog is-rounded"
    id="modal-post"
    onClick={[Function]}
  >
    <span />
  </dialog>
</div>
`;

exports[`React Test CommentRow Should render correctly 1`] = `
<section
  className="message -left"
>
  <i
    className="nes-bcrikko icon"
  />
  <div
    className="nes-balloon from-left"
  >
    <p />
  </div>
</section>
`;

exports[`React Test Post Should render correctly 1`] = `
<div
  className="post-container"
>
  <p
    className="user nes-pointer"
    onClick={[Function]}
  >
    <i
      className="nes-ash"
    />
     
    <span>
       
      test name
       
    </span>
  </p>
  <p
    className="title"
  >
     
    sunt aut facere repellat provident occaecati excepturi optio reprehenderit
     
  </p>
  <p
    className="body"
  >
     
    lorem
     
  </p>
  <span
    className="comments-link nes-pointer"
    onClick={[Function]}
  >
     Load Comments 
  </span>
</div>
`;

exports[`React Test PostModal Should render correctly 1`] = `
<dialog
  className="nes-dialog is-rounded"
  id="modal-post"
  onClick={[MockFunction]}
>
  <span />
</dialog>
`;
